### PR TITLE
Reference scripts impl - publication of a script

### PR DESCRIPTION
**Pull request by paweljakubas**
- [x] logic in `constructTransaction` on api and `Cardano.Wallet` level
- [x] extension of `TransactionCtx`
- [x] use the added field in `mkUnsignedTx`
- [x] fix `toCardanoTxOut`
- [x] adjust fee when reference script is present
- [x] integration test showing reference script publication 

### Comments

PR tackles publication of reference script to be used in later transactions by using reference inputs when minting/burning.
Script is propagated in blockchain in the first of TxOut of a given transaction. It is shown in integration tests that construction/signing/submitting of such a transaction works, plus when decoding transaction a script is present in witness count as expected.  

### Issue Number
https://cardanofoundation.atlassian.net/browse/ADP-3090

### CODE REVIEW

File: lib/wallet/src/Cardano/Wallet/Write/Tx/Balance.hs [hunk 1]
```diff
@@ -901,6 +906,8 @@ selectAssets era (ProtocolParameters pp) utxoAssumptions outs redeemers
                 , txChange = view #skeletonChange skeleton
                 , txPaymentTemplate = view #template <$>
                     assumedInputScriptTemplate utxoAssumptions
+                , txReferenceScript =
```
Comment (Anviking) [inline lib/wallet/src/Cardano/Wallet/Write/Tx/Balance.hs]: As said in DM: The changes to `Write.{Balance, SizeEstimation, UTxOAssumptions}` shouldn't be needed. As long as the new inline-script-containing `TxOut` is part of the partial tx given to `balanceTx`, then `balanceTx` will know of its size and fee-contribution. So it should _just work_ with the changes reverted.

Comment (paweljakubas) [reply lib/wallet/src/Cardano/Wallet/Write/Tx/Balance.hs]: thanks! indeed adjusting `estimateTxSize` was not needed!

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs [hunk 2]
```diff
@@ -2623,6 +2633,43 @@ constructTransaction api argGenChange knownPools poolStatus apiWalletId body = d
...
+        guardWrongScriptTemplate
+            :: Script Cosigner -> Either ErrConstructTx ()
+        guardWrongScriptTemplate apiScript =
+            when (wrongMintingTemplate apiScript)
+                $ Left ErrConstructTxWrongMintingBurningTemplate
+          where
+            wrongMintingTemplate script =
+                isLeft (validateScriptOfTemplate RecommendedValidation script)
+                || countCosigners script /= (1 :: Int)
+                || existsNonZeroCosigner script
```
Comment (HeinrichApfelmus) [inline lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs:2653]: Why are these conditions checked? According to the specification (swagger.yaml), any script template should be acceptable. 🤔

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs [hunk 3]
```diff
@@ -2623,6 +2633,43 @@ constructTransaction api argGenChange knownPools poolStatus apiWalletId body = d
...
+            :: (SlotNo, SlotNo)
+            -> Script Cosigner
+            -> Either ErrConstructTx ()
+        guardOutsideValidityInterval (before, hereafter) script =
+            when notWithinValidityInterval $
+                Left ErrConstructTxValidityIntervalNotWithinScriptTimelock
+          where
+            notWithinValidityInterval =
+                not $ withinSlotInterval before hereafter $
+                    scriptSlotIntervals script
```
Comment (HeinrichApfelmus) [inline lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs]: Same question as above, with the following question added: As the script is supposed to be used as a reference input, it would be ok for the script to have a much larger validity interval (`scriptSlotIntervals script`) than the transaction which is used to put the script on chain (`(before, herafter)`)? I.e. I believe that `guardOutsideValidityInterval` should not be present. 🤔

**Review by HeinrichApfelmus — CHANGES_REQUESTED** [review]
The implementation and test look sensible to me, but it appears to me that the validation checks when parsing a reference script template do not match the specification. 🤔

(In addition, I believe that the specific validation checks here should not be performed — definitely for the validity interval one; less definitely for the checks on the number of cosigners, as there is some correlation to which minting scripts that we allow.)

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs (hunk 2 above)
Comment (paweljakubas) [reply lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs:2653]: script template allows more than one cosigner in general. But here we are using script for minting/burning in SHELLEY (not in shared style) so we need to make sure one cosigner is present.

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs (hunk 3 above)
Comment (paweljakubas) [reply lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs]: there is ledger rule that requires script timelocks determined by (hereafter, before) to encapsulate validity interval. If not the case ledger will not accept this. Ask Jared why is that way ;-)

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs (hunk 2 above)
Comment (HeinrichApfelmus) [reply lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs:2653]: That's alright with me — all I'm asking for is that this fact is highlighted in the specification. 🤓📓 Could you amend the documents [mint-burn.md][1] and [swagger.yaml][2] to describe the allowed values for `reference_policy_script_template`?

(By definition, if the implementation is different from the specification, that's a bug.)

  [1]: https://github.com/cardano-foundation/cardano-wallet/blob/master/specifications/api/mint-burn.md
  [2]: https://github.com/cardano-foundation/cardano-wallet/blob/master/specifications/api/swagger.yaml

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs (hunk 3 above)
Comment (HeinrichApfelmus) [reply lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs]: Which ledger rule are you referring to specifically? 🤔 

The UTXOW rule from the Babbage era does require that all reference scripts that are used in the transaction fall within the validity interval:

validateScript

However, this does *not* entail that the scripts which are put in the `txouts` need to fall within the validity interval of the transaction — they are different in that they not used by the transaction to mint or burn assets; rather, they are put on chain for future use.

In other words, the validity interval can only affect the validity of transactions that *use* the reference script, not the validity of transactions that *create* the reference script.

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs (hunk 2 above)
Comment (paweljakubas) [reply lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs:2653]: ah, ok. done

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs (hunk 3 above)
Comment (paweljakubas) [reply lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs]: please see the following integration test https://github.com/cardano-foundation/cardano-wallet/blob/master/lib/wallet/integration/src/Test/Integration/Scenario/API/Shelley/TransactionsNew.hs#L3600 . Generally speaking we need script's timelock interval be in validity interval. If it is not the case, as in the test above node will end up with error upon submitting. This is checked somewhere in ledger code (not in UTXOW rules specified unfortunately)

Comment (HeinrichApfelmus) [reply lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs]: But as I tried to convey, the integration test `TRANS_NEW_CREATE_10e` is concerned with the *use* of the script, i.e. the `policy_script_template` parameter. This is indeed checked by the UTXOW rule.

I believe that the node will not reject the script in the integration test if you put it in the `reference_policy_script_template`. I can check tomorrow if you like.

Comment (paweljakubas) [reply lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs]: you are right! when script is published to be used later it does not make any sense to check against validity as we do not know when it is going to be used (we know only that later). Moreover we cannot use the published script in the same transaction.

Thanks @HeinrichApfelmus for raising this very subtle thing! Good catch!

I have removed this validation!

File: specifications/api/mint-burn.md [hunk 4]
```diff
@@ -16,7 +16,7 @@ Specifically:
 
 1. Creation of a transaction output that contains a minting script and is suitable for use as reference input.
 
-    In the `reference_policy_script_template` field, you can optionally add a script template. The HTTP endpoint will map this script template into a script using the wallet's policy public key, and this script will be included in the first transaction output (i.e. at index `0`) of the transaction.
+    In the `reference_policy_script_template` field, you can optionally add a script template. The HTTP endpoint will map this script template into a script using the wallet's policy public key, and this script will be included in the first transaction output (i.e. at index `0`) of the transaction. It is worth mentioning that for shelley wallets script template needs to include one cosigner only.
```
Comment (HeinrichApfelmus) [inline specifications/api/mint-burn.md]:
```suggestion
    In the `reference_policy_script_template` field, you can optionally add a script template. The HTTP endpoint will map this script template into a script using the wallet's policy public key, and this script will be included in the first transaction output (i.e. at index `0`) of the transaction. For Shelley-style wallets, the script template must contain a single cosigner only, but it may include time locks.
```

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs (hunk 2 above)
Comment (HeinrichApfelmus) [reply lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs:2653]: Thanks for adding a description of the restriction to [mint-burn.md](https://github.com/cardano-foundation/cardano-wallet/blob/master/specifications/api/mint-burn.md) — could you add it to [swagger.yaml](https://github.com/cardano-foundation/cardano-wallet/blob/master/specifications/api/swagger.yaml) as well?

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs (hunk 3 above)
Comment (HeinrichApfelmus) [reply lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs]: 🤓 Thanks!

**Review by HeinrichApfelmus — APPROVED** [review]
Thanks! 😊 No major objections anymore. Could you address the remaining comments? — Small change to description in mint-burn.md, and I would like to see a change to swagger.yaml as well.

File: specifications/api/mint-burn.md (hunk 4 above)
Comment (paweljakubas) [reply specifications/api/mint-burn.md]: done
